### ISSUE TITLE
[Android] Is it possible to write data from one app, and read from another?

### ISSUE THREAD

**Issue by christian-hawk** [mogol/flutter_secure_storage#490]
I would like to implement [OpenID Connect Native SSO for Mobile Apps 1.0](https://openid.net/specs/openid-connect-native-sso-1_0.html) flow, which requires to "App 1" to *store* 2 keys/values (`id_token` and `device_secret`). Those values should only be accessible to "App 2"

"App 2" will read `id_token` and `device_secret` and exchange them for a token.

What would be the recommended approach, if possible to do it with `flutter_secure_storage`?

Thanks

**Comment by nynymike** [comment]
For example, could a Flutter app store secrets in something like the iCloud Keystore, where the same vendor can share a secret between two different apps?

**Comment by subzero911** [comment]
Hi @christian-hawk , i'm doing a research on a same task. In short - no, this plugin is not suitable, it can keep the data within one app only.
The shared storage for credentials on Android is AccountManager https://developer.android.com/reference/android/accounts/AccountManager
Not sure about iOS, probably Keychain can be suitable.
There's no ready-made SSO package on pub.dev. You'll have to write your own native plugin for saving/retrieving data, and a Dart wrapper for it.

**Comment by christian-hawk** [comment]
Hey @subzero911 thank you very much for the input. Are you in the task of implementing [OpenID Connect Native SSO for Mobile Apps 1.0](https://openid.net/specs/openid-connect-native-sso-1_0.html) or just sharing data between apps?

The reason I ask, I've checked [Account Manager] (https://developer.android.com/reference/android/accounts/AccountManager) link you provided, I does not look like it's ready to Native SSO Flow (storing id_token and device_id from one app, fetch it from another). Am I wrong? Wouldn't `Keystore` be a better fit, using `sharedPreferences`?

**Comment by subzero911** [comment]
@christian-hawk, SSO is implemented on backend.
On mobile app side, you just have to keep all your accounts in some shared zone, like AccountManager (not sure about Keystore, and if it's good for this task). sharedPreferences and flutter_secure_storage are definitely not suitable, as they keep only app-wide data.

When you authorize in your app (via WebView or REST service - it depends on your backend), you receive an access and refresh token pair, and add them into AccountManager and also save them into local storage (sharedPreferences, HIve).
When you reopen your app, you can check - if you have a data in a local storage, you can silently log in.
If you try to login, but your access token is expired, and you can't renew it - then you load all accounts from your AccountManager and show a modal bottom sheet with account list to user to choose from.
If you don't have a local data - the same, you prompt a user to choose an account.
When you log off - you erase the local data, and prompt a user to choose an account.
There also should be a "I am a new user" button in a bottom sheet to login / signup as a new user.
Consider also having a "Delete local account" button to remove an account from AccountManager (it won't delete your account on server, you can enter the same credentials and add it back to AccountManager).

So basically SSO is just a convenience feature to not to enter your credentials every time you are log in, but having an OS-wide accounts list, which can be used by several apps.

**Comment by Mik77o** [comment]
We have a similar issue with: https://github.com/aws-amplify/amplify-flutter/issues/4645. While iOS uses Keychain, there is no such option on Android (not implemented there).

**Comment by juliansteenbakker** [comment]
I am closing all older issues. If this issue still exists in the latest version, please let me know.
